File: source/gui/hierarchy_gui.cpp
#include "../../include/gui/hierarchy_gui.h"
#include "../../lib/imgui/imgui.h"
#include "../../include/core/engine.h"
#include "../../include/core/console.h"

#include <string>

void HierarchyGUI::render(int screenW, int screenH) {
	ImGui::Begin("Hierarchy");

	//ImGui::Text("%s", "Hierarchy");
	//char textBuffer[256] = {'\0'};

	//ImGui::InputText("Search", textBuffer, sizeof(textBuffer));
	ImGui::Separator();

	//if (Engine::objects.size() > 0) {
	//	/*if (ImGui::TreeNodeEx(Engine::objects[0].name.c_str(), ImGuiTreeNodeFlags_DefaultOpen)) {
	//		if (ImGui::TreeNodeEx(Engine::objects[1].name.c_str(), ImGuiTreeNodeFlags_DefaultOpen)) {
	//			if (ImGui::TreeNodeEx(Engine::objects[2].name.c_str(), ImGuiTreeNodeFlags_Leaf)) {

	//				ImGui::TreePop();
	//			}
	//			ImGui::TreePop();
	//		}
	//		ImGui::TreePop();
	//	}*/
	//	//bool nodeOpen = false;
	//	//ImGuiTreeNodeFlags nodeFlags = ImGuiTreeNodeFlags_DefaultOpen;
	//	for (int i = 0; i < Engine::objects.size(); ++i)
	//	{
	//		EngineObject obj = Engine::objects[i];
	//		/*int prevIndex = i - 1;
	//		if (prevIndex < 0)
	//			prevIndex = 0;*/

	//		if (!obj.transform.has_parent())
	//			draw_nodes(obj, obj.transform.children);

	//		/*Console::write("OBJECT: [" + obj.name + "]");
	//		for (int i = 0; i < obj.transform.child_count(); ++i)
	//		{
	//			Console::write("[Child: " + std::to_string(i) + "]	" + obj.transform.children[i].engineObject->name);
	//		}*/
	//	}
	//}

	//delete[] textBuffer;
	ImGui::End();
}
//void HierarchyGUI::draw_nodes(EngineObject& obj, std::vector<Transform> children) {
//	ImGuiTreeNodeFlags nodeFlags = children.size() > 0 ? ImGuiTreeNodeFlags_DefaultOpen : ImGuiTreeNodeFlags_Leaf;
//	//ImGuiTreeNodeFlags nodeFlags = ImGuiTreeNodeFlags_DefaultOpen;
//	if (ImGui::TreeNodeEx(obj.name.c_str(), nodeFlags)) {
//		for (const Transform t : children)
//		{
//			draw_nodes(*t.engineObject, t.children);
//		}
//
//		ImGui::TreePop();
//	}
//}
ImGuiInputTextCallback HierarchyGUI::search_callback() {
	return nullptr;
}

File: GUI/Windows/HierarchyGUI.cs
using AssemblyEngine.SceneManagement;
using Hexa.NET.ImGui;

namespace AssemblyEngine.GUI
{
    internal class HierarchyGUI : GUIWindow
    {
        private string renameBuffer;
        public override void Draw()
        {
            ImGui.Begin("Hierarchy");
            ImGui.Separator();

            ImGuiTreeNodeFlags nodeFlags = ImGuiTreeNodeFlags.DefaultOpen | ImGuiTreeNodeFlags.OpenOnArrow;

            foreach (var scene in SceneManager.loadedScenes)
            {
                if (ImGui.TreeNodeEx(scene.name, nodeFlags))
                {
                    foreach (var engineObject in scene.engineObjects)
                    {
                        if (engineObject.transform.parent == null)
                        {
                            DrawNodes(engineObject);
                        }
                    }

                    ImGui.TreePop();
                }
            }

            if (ImGui.IsWindowHovered() && ImGui.IsMouseClicked(ImGuiMouseButton.Right))
            {
                ImGui.OpenPopup("Create Object");
            }
            if (ImGui.BeginPopup("Create Object"))
            {
                ImGui.Text("Create Object");
                if (ImGui.MenuItem("Empty"))
                {
                    EngineObjectFactory.Instantiate("New Object");
                }
                if (ImGui.BeginMenu("3D"))
                {
                    if (ImGui.MenuItem("Cube"))
                    {

                    }
                    if (ImGui.MenuItem("Inverted Cube"))
                    {

                    }
                    if (ImGui.MenuItem("Plane"))
                    {

                    }
                    ImGui.EndMenu();
                }
                ImGui.EndPopup();
            }

            ImGui.End();
        }
        private void DrawNodes(EngineObject engineObject)
        {
            ImGui.PushID(engineObject.objectGuid.ToString());

            ImGuiTreeNodeFlags nodeFlags = engineObject.transform.children.Count > 0 ? ImGuiTreeNodeFlags.DefaultOpen : ImGuiTreeNodeFlags.Leaf;

            nodeFlags |= engineObject == EngineItemSelector.selectedObject ? ImGuiTreeNodeFlags.Selected : ImGuiTreeNodeFlags.None;

            nodeFlags |= ImGuiTreeNodeFlags.OpenOnArrow;
            
            if (ImGui.TreeNodeEx(engineObject.name, nodeFlags))
            {
                if (ImGui.IsItemClicked())
                {
                    EngineItemSelector.selectedObject = engineObject;
                }
                if (ImGui.IsItemHovered() && ImGui.IsMouseDoubleClicked(ImGuiMouseButton.Left))
                {
                    EngineItemSelector.objectToRename = engineObject;
                    renameBuffer = engineObject.name;
                }
                if (engineObject == EngineItemSelector.objectToRename)
                {
                    ImGuiInputTextFlags inputFlags = ImGuiInputTextFlags.EnterReturnsTrue | ImGuiInputTextFlags.AutoSelectAll;
                    ImGui.SetNextItemAllowOverlap();
                    if (ImGui.InputText("##rename", ref renameBuffer, 128, inputFlags))
                    {
                        EngineItemSelector.objectToRename = null;

                        if (renameBuffer != string.Empty)
                            engineObject.name = renameBuffer;
                    }
                }
                foreach (Transform child in engineObject.transform.children)
                {
                    DrawNodes(child.engineObject);
                }
            
            	ImGui.TreePop();
            }

            ImGui.PopID();
        }
    }
}